File: backend/cubebox/api/schemas/conversation_search.py
"""Pydantic schemas for the conversation search route."""

from pydantic import BaseModel, Field


class SearchResultSchema(BaseModel):
    conversation_id: str
    title: str
    snippet: str
    match_offsets: list[tuple[int, int]] = Field(default_factory=list)
    matched_message_seq: int | None = None
    matched_at: str | None = None
    score: float


class SearchResponseSchema(BaseModel):
    results: list[SearchResultSchema]
    lexical_count: int
    vector_count: int
    fused_count: int

File: frontend/packages/core/src/api/conversation-search.ts
import { toApiError, type ApiClient } from './client'

export interface SearchResult {
  conversation_id: string
  title: string
  snippet: string
  match_offsets: [number, number][]
  matched_message_seq: number | null
  matched_at: string | null
  score: number
}

export interface SearchResponse {
  results: SearchResult[]
  lexical_count: number
  vector_count: number
  fused_count: number
}

export async function searchConversations(
  client: ApiClient,
  q: string,
  limit = 8,
): Promise<SearchResponse> {
  // ApiClient.get returns Promise<Response>; it auto-injects the `/ws/{id}/`
  // segment via injectWorkspace, so we pass the scoped suffix only.
  const path = `/api/v1/conversations/search?q=${encodeURIComponent(q)}&limit=${limit}`
  const res = await client.get(path)
  if (!res.ok) throw await toApiError(res)
  return (await res.json()) as SearchResponse
}
